### ISSUE TITLE
Detection of `system` switch

### ISSUE THREAD

**Issue by thranur** [ocaml/opam#2452]
I'd like to try a locally compiled version of OCaml that I installed in a directory that is not in the PATH, so I tried the following:

    export PATH=/path/to/locally/installed/ocaml/bin:$PATH
    opam switch system

But this results in:

    [ERROR] No OCaml compiler found in path. You should use:

            opam init --comp=VERSION

If I type something like `ocamlc --version`, or `ocamlc.opt --version`, all of them work and return the locally installed version (`4.03.0+beta1`).

If I type `opam config env`, the reported value for `PATH` does include the directory I just added, so in theory the commands `ocaml`, `ocamlc`, etc. should work.

Is the mechanism used to detect the `system` compiler based on something other than `PATH`? If so, would it be useful to report more information about where OPAM is searching?

I also tried `opam switch system --debug` but it didn't produce any useful information.

**Comment by AltGr** [comment]
That's quite strange indeed; opam only uses the PATH, and the shell to look for `ocamlc`. One reason that I could imagine is that opam internally changes the PATH to avoid mixing the contents of different switches ; but what it does is only remove components starting with `$OPAMROOT`.

**Comment by thranur** [comment]
Oh, OK, I hadn't realized that the `system` switch is defined only at `opam init` time, and if there's no OCaml compiler in the PATH, it won't show any `system` switch, even if OCaml is later added to the PATH (which makes sense, otherwise the currently configured OPAM switch would always show up as `system`).

I just did `opam init --root=/tmp/test` and it showed up the currently opam-based OCaml as being "system", even though I had just installed my Ubuntu's `ocaml` package. At first I had assumed that it was somewhat hardcoded based on the installation path (e.g. `/usr/bin` or `/usr/local/bin`).

The error message does mention `opam init`, but I was focused on the first line ("No OCaml compiler found in path."). The only possible suggestion I'd make would be to add something like "No OCaml compiler found in path *during 'opam init'*", but I'm sure there are other situations that would cause this error message to appear, so it's not ideal either.

Anyway, thanks for the comment, it helped me track it out.

**Comment by myfreeweb** [comment]
With opam-2.0.0-rc3, looks like `opam init` doesn't like the ocamlc that's on my PATH, and stubbornly starts downloading `ocaml-base-compiler` 4.07.0 :(

Even though `opam config list` shows that OPAM clearly does in fact see my ocamlc:

```
sys-ocaml-version 4.08.0+dev0-2018-04-09 # OCaml version present on your system independently of opam, if any
```

(I *can't* let it build 4.07.0, that wouldn't work since I've [just ported OCaml itself to this platform](https://github.com/ocaml/ocaml/pull/1904))

**Comment by dra27** [comment]
@myfreeweb - this isn't really related to this issue (this issue was with OPAM 1.2.2). it'd be better to open a new one if you want to continue the discussion further.

However, the problem is that you're using a development compiler as a system compiler, which is not expected usage - it means that opam doesn't find any version of `ocaml-system` to install (it's nothing to do with `PATH`). You'll need to fork the opam-repository repo's 2.0.0 branch and add an `ocaml-system.4.08.0` directory to `packages/ocaml-system` (you can use the `ocaml-system.4.06.1` directory as a template). You will just need to change the `available` field of the `opam` file in `available: sys-ocaml-version = "4.06.1"` to match the version number of your system compiler (I'd suggest using a range, given that you're using a development compiler)
